Value incorrectly flagged as being possibly None after `or` predicate ensures it isn't None

Invoked with: ` python3.6 -m mypy --warn-no-return --strict-optional test_or_shortcut.py`

```python
from typing import Optional, Any


def test_or_shortcut(value: Optional[Any]) -> None:
    if not value:
        pass
    if not value or value.get('foo') == 'hello':
        pass
```
Fails with:
```
test_or_shortcut.py:7: error: Item "None" of "Optional[Any]" has no attribute "get"
```

Oddly if I remove the first branch things typecheck fine:

```python
from typing import Optional, Any


def test_or_shortcut(value: Optional[Any]) -> None:
    if not value or value.get('foo') == 'hello':
        pass
```

--- a/mypy/types.py
+++ b/mypy/types.py
@@ -1699,6 +1699,10 @@ def true_or_false(t: Type) -> Type:
     """
     Unrestricted version of t with both True-ish and False-ish values
     """
+    if isinstance(t, UnionType):
+        new_items = [true_or_false(item) for item in t.items]
+        return UnionType.make_simplified_union(new_items, line=t.line, column=t.column)
+
     new_t = copy_type(t)
     new_t.can_be_true = type(new_t).can_be_true
     new_t.can_be_false = type(new_t).can_be_false

--- a/test-data/unit/check-optional.test
+++ b/test-data/unit/check-optional.test
@@ -594,3 +594,12 @@ x: Optional[Union[int, str]]
 reveal_type(x) # E: Revealed type is 'Union[builtins.int, builtins.str, builtins.None]'
 y: Optional[Union[int, None]]
 reveal_type(y) # E: Revealed type is 'Union[builtins.int, builtins.None]'
+
+[case testUnionTruthinessTracking]
+from typing import Optional, Any
+def test_or_shortcut(value: Optional[Any]) -> None:
+    if not value:
+        pass
+    if not value or value.get('foo') == 'hello':
+        pass
+[builtins fixtures/bool.pyi]

